rust/yarp/build.rs

```diff
@@ -0,0 +1,507 @@
+use serde::Deserialize;
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+
+#[derive(Debug, Deserialize)]
+enum NodeParamType {
+    #[serde(rename = "node")]
+    Node,
+
+    #[serde(rename = "node?")]
+    OptionalNode,
+
+    #[serde(rename = "node[]")]
+    NodeList,
+
+    #[serde(rename = "string")]
+    String,
+
+    #[serde(rename = "constant")]
+    Constant,
+
+    #[serde(rename = "constant[]")]
+    ConstantList,
+
+    #[serde(rename = "location")]
+    Location,
+
+    #[serde(rename = "location?")]
+    OptionalLocation,
+
+    #[serde(rename = "location[]")]
+    LocationList,
+
+    #[serde(rename = "uint32")]
+    UInt32,
+
+    #[serde(rename = "flags")]
+    Flags
+}
+
+#[derive(Debug, Deserialize)]
+struct NodeParam {
+    name: String,
+    r#type: NodeParamType,
+    kind: Option<String>
+}
+
+#[derive(Debug, Deserialize)]
+struct Node {
+    name: String,
+    child_nodes: Option<Vec<NodeParam>>,
+    comment: String
+}
+
+#[derive(Debug, Deserialize)]
+struct Config {
+    nodes: Vec<Node>
+}
+
+/// The main function for the build script. This will be run by Cargo when
+/// building the library.
+/// 
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let file = std::fs::File::open(config_path())?;
+    let config: Config = serde_yaml::from_reader(file)?;
+
+    write_bindings(config)?;
+    Ok(())
+}
+
+/// Gets the path to the config.yml file at `[root]/config.yml`.
+///
+fn config_path() -> PathBuf {
+    cargo_manifest_path()
+        .join("../../config.yml")
+        .canonicalize()
+        .unwrap()
+}
+
+/// Gets the path to the root directory at `[root]`.
+/// 
+fn cargo_manifest_path() -> PathBuf {
+    PathBuf::from(std::env::var_os("CARGO_MANIFEST_DIR").unwrap())
+}
+
+/// Returns the name of a C struct from the given node name.
+fn struct_name(name: &str) -> String {
+    let mut result = String::with_capacity(1 + name.len());
+
+    for char in name.chars() {
+        if char.is_uppercase() {
+            result.push('_');
+        }
+        result.push(char.to_lowercase().next().unwrap());
+    }
+
+    result
+}
+
+/// Returns the name of the C type from the given node name.
+fn type_name(name: &str) -> String {
+    let mut result = String::with_capacity(8 + name.len());
+    result.push_str("YP_NODE");
+
+    for char in name.chars() {
+        if char.is_uppercase() {
+            result.push('_');
+        }
+        result.push(char.to_uppercase().next().unwrap());
+    }
+
+    result
+}
+
+/// Write the generated struct for the node to the file.
+/// 
+fn write_node(file: &mut File, node: &Node) -> Result<(), Box<dyn std::error::Error>> {
```

Comment:
Any interesting in using [quote](https://docs.rs/quote/latest/quote/) for generating this code? I once had a crate where I generated a bunch of Rust code using handlebars and that got to be pretty annoying to maintain.

Reply:
I attempted to use `quote` but it just returns a token stream, and getting it back to nicely formatted source appears to not be an option.

Reply:
You can use the `prettyplease` crate for that